dynamiqs/qarrays/sparse_qarray.py

```diff
@@ -275,6 +275,39 @@ def __rmatmul__(self, other: Array) -> Array:
 
         return NotImplemented
 
+    def _tensor_dense(self, other: ArrayLike) -> Array:
+        N = self.diags.shape[-1]
+        p, q = other.shape
+        out = jnp.empty((N * p, N * q), dtype=jnp.int32)
+        for self_offset, self_diag in zip(self.offsets, self.diags):
+            i = jnp.arange(N + 1 - self_offset)
+            j = jnp.arange(self_offset, N + 1)
+            start = max(0, self_offset)
+            end = min(N, N + self_offset)
+            for k, d in enumerate(self_diag[start:end]):
+                out = jax.lax.dynamic_update_slice(out, d * other, (N * i[k], N * j[k]))
+        return out
```

Comment:
If the output is a dense array, I would suggest converting the sparse array to a dense one (with `.to_dense()`) and calling the `__and__` of `DenseQArray`. My guess is that it should be almost as fast.